mathics/algorithm/parts.py

```diff
@@ -21,78 +24,99 @@
+            raise PartDepthError(indices[0])
+
+        pos = indices[0]
+        elements = expression.elements
+        try:
+            if pos > 0:
+                part = elements[pos - 1]
+            elif pos == 0:
+                part = expression.get_head()
+            else:
+                part = elements[pos]
+        except IndexError:
+            raise PartRangeError
+        return get_subpart(part, indices[1:])
 
+    return get_subpart(expression, indices).copy()
 
-def set_part(varlist, indices, newval):
-    "Simple part replacement. indices must be a list of python integers."
 
-    def rec(cur, rest):
-        if len(rest) > 1:
-            pos = rest[0]
-            if isinstance(cur, Atom):
+def set_part(expression, indices: List[int], newval) -> BaseElement:
+    """Replace all parts of ``expression`` specified by ``indicies`` with the
+    ``newval`` value. Return the modified compound expression.
+    """
+
+    def set_subpart(expression, indices: List[int]) -> BaseElement:
+        """Recursive work-horse portion of ``set_part()`` that replaces pieces
+        of ``expression`` with outer variable ``newval`` based on ``indices``.
+        ``expression`` and ``indices`` are smaller parts of the corresponding variables at the outer level.
+        """
+        if len(indices) > 1:
+            pos = indices[0]
+            if isinstance(expression, Atom):
                 raise PartDepthError
             try:
                 if pos > 0:
-                    part = cur.elements[pos - 1]
+                    part = expression.elements[pos - 1]
                 elif pos == 0:
-                    part = cur.get_head()
+                    part = expression.get_head()
                 else:
-                    part = cur.elements[pos]
+                    part = expression.elements[pos]
             except IndexError:
                 raise PartRangeError
-            return rec(part, rest[1:])
-        elif len(rest) == 1:
-            pos = rest[0]
-            if isinstance(cur, Atom):
+            set_subpart(part, indices[1:])
+            return expression
+        elif len(indices) == 1:
+            pos = indices[0]
+            if isinstance(expression, Atom):
                 raise PartDepthError
             try:
                 if pos > 0:
-                    cur.set_element(pos - 1, newval)
+                    expression.set_element(pos - 1, newval)
                 elif pos == 0:
-                    cur.set_head(newval)
+                    # We may have to replace the entire ``expression``
+                    # variable when changing position 0 or Head. This
+                    # happens when the before and after are
+                    # class objects are different.
+
+                    # Right now, we need to only worry about
+                    # converting between ``Expression`` and
+                    # ``ListExpression`` or vice vera.  In the code
```

Comment:
I would think that if you change the head, you would like to build a new expression of the generic type `Expression`. Suppose for instance that you have a `List`, and then you want to change the head to be `Raster`. A simple way to do this is to do `Expression(SymbolRaster, *cur._elements, cur.elements_properties)`. Then, in the next evaluation, if `Raster` has associated a  `BuiltinElement` object, the expression can be converted into a `Raster(*elements, cur.elements_properties)`

Reply:
In my opinion change the Head element is kind of a dodgy thing to do. 

Overall though if you want a performant interpreter, we need to give up on the notion that you there are several equivalent forms that a compound expression can be in and the interpreter is silently checking and converting forms. 

Right now, it is true that we need to be able to handle Expression(SymbolList, ...) as well as ListExpression(...) but this is transitory in the same way that you could write Expression("Plus", 1, 2) or Expression(SymbolPlus, Integer1, Integer2) and have these be interchangeable. 

In the future though I don't think we can or should allow this. In my view you are overthinking and over complicating things from the outset. Let the complexity come to you rather than seek it out. 

Related but in my view very very different is something like the recently changed addition where you can now take a ListExpression and index into it eg. `list_expression[0]` for the (non-head) first element. At some point I was going to also add the ability to do `len(list_expression)`. And we have this feature ware you can write `Integer0 + Integer1` which is equivalent to `Expression(Symbol Plus, Integer0, Integer1)` .  

These are different things, because in the latter, everything really is strongly typed and there _aren't_ conversions and checking going on: the `+` and `len()` are more syntactic conveniences but underneath the types that they work on are not malleable.